pkg/webui/oauth/views/change-password/index.js

```diff
@@ -0,0 +1,157 @@
+import { defineMessages } from 'react-intl'
+import { push, replace } from 'connected-react-router'
+import { connect } from 'react-redux'
+import * as Yup from 'yup'
+
+import api from '../../api'
+import sharedMessages from '../../../lib/shared-messages'
+import { digit, uppercase, special } from '../../lib/regexp'
+
+import Button from '../../../components/button'
+import Form from '../../../components/form'
+import Input from '../../../components/input'
+import SubmitButton from '../../../components/submit-button'
+import IntlHelmet from '../../../lib/components/intl-helmet'
+import Message from '../../../lib/components/message'
+
+import style from '../create-account/create-account.styl'
+
+const m = defineMessages({
+  changePassword: 'Change Password',
+  passwordChanged: 'Password changed successfully',
+})
+
+const validationSchema = Yup.object().shape({
+  new_password: Yup.string()
+    .min(8)
+    .matches(digit, sharedMessages.validatePasswordDigit)
+    .matches(uppercase, sharedMessages.validatePasswordUppercase)
+    .matches(special, sharedMessages.validatePasswordSpecial)
+    .required(sharedMessages.validateRequired),
+  new_confirm: Yup.string()
+    .oneOf([ Yup.ref('new_password'), null ], sharedMessages.validatePasswordMatch)
+    .min(8)
+    .required(sharedMessages.validateRequired),
+})
+
+const initialValues = {
+  new_password: '',
+  new_confirm: '',
+}
+
+@connect(null, {
+  push,
+  replace,
+})
+@bind
+export default class ChangePassword extends React.PureComponent {
+  constructor (props) {
+    super(props)
+    this.state = {
+      error: '',
+      info: '',
+    }
+  }
+
+  async handleSubmit (values, { setSubmitting }) {
+    const { match: { params }, push } = this.props
+
+    try {
+      await api.users.update_password( params.user_id, {
+        user_ids: { user_id: params.user_id },
+        new: values.new_password,
+        old: params.token,
+      })
+
+      push('/oauth/login', {
+        info: m.passwordChanged,
+      })
+    } catch (error) {
+      this.setState({
+        error: error.response.data,
+        info: '',
+      })
+    } finally {
+      setSubmitting(false)
+    }
+  }
+
+  navigateToLogin () {
+    const { replace } = this.props
+
+    replace('/oauth/login')
+  }
+
+  render () {
+    const { error, info } = this.state
+    const { match: { params }} = this.props
+
+    return (
+      <Container className={style.fullHeight}>
+        <Row justify="center" align="center" className={style.fullHeight}>
+          <Col sm={12} md={8} lg={5}>
+            <IntlHelmet title={m.changePassword} />
+            <Message content={m.changePassword} component="h1" className={style.title} />
+            <Form
+              onSubmit={this.handleSubmit}
+              initialValues={initialValues}
+              error={error}
+              info={info}
+              validationSchema={validationSchema}
+              horizontal={false}
+            >
+              <Form.Field
```

Comment:
Missing a form field for old password here (can be prefilled from query).

Reply:
> What do you think about a flow where, after requesting the temporary password, the web UI would immediately switch to change-password?

This would not work if we intend to supply the `temporary_password` in URI, since the `temporary_password` would be available only in the e-mail. If we redirect to change-password view there would be no `temporary_password` supplied.

> Missing a form field for old password here (can be prefilled from query).
This should be there for the case when a logged in user wants to change the password.

Reply:
> > Missing a form field for old password here (can be prefilled from query).
> This should be there for the case when a logged in user wants to change the password.

The temporary password is equivalent to the current password here, so why not treat them the same?